flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/sink/shuffle/TestMapRangePartitioner.java

```diff
@@ -0,0 +1,511 @@
+            SORT_KEY_2,
+            new MapRangePartitioner.KeyAssignment(new int[] {6, 7}, new long[] {5L, 7L}),
+            SORT_KEY_3,
+            new MapRangePartitioner.KeyAssignment(new int[] {7, 8}, new long[] {2L, 2L}),
+            SORT_KEY_4,
+            new MapRangePartitioner.KeyAssignment(new int[] {8}, new long[] {1L}),
+            SORT_KEY_5,
+            new MapRangePartitioner.KeyAssignment(new int[] {8}, new long[] {1L}),
+            SORT_KEY_6,
+            new MapRangePartitioner.KeyAssignment(new int[] {8}, new long[] {1L}),
+            SORT_KEY_7,
+            new MapRangePartitioner.KeyAssignment(new int[] {8}, new long[] {1L}),
+            SORT_KEY_8,
+            new MapRangePartitioner.KeyAssignment(new int[] {8}, new long[] {1L}),
+            SORT_KEY_9,
+            new MapRangePartitioner.KeyAssignment(new int[] {8}, new long[] {1L}));
+    Map<SortKey, MapRangePartitioner.KeyAssignment> actualAssignment =
+        partitioner.assignment(numPartitions);
+    Assertions.assertThat(actualAssignment).isEqualTo(expectedAssignment);
+
+    // key: subtask id
+    // value pair: first is the assigned weight, second is the number of assigned keys
+    Map<Integer, Pair<Long, Integer>> expectedAssignmentInfo =
+        ImmutableMap.of(
+            0,
+            Pair.of(9L, 1),
+            1,
+            Pair.of(9L, 1),
+            2,
+            Pair.of(9L, 1),
+            3,
+            Pair.of(9L, 2),
+            4,
+            Pair.of(9L, 1),
+            5,
+            Pair.of(9L, 1),
+            6,
+            Pair.of(9L, 2),
+            7,
+            Pair.of(9L, 2),
+            8,
+            Pair.of(8L, 7));
+    Map<Integer, Pair<Long, Integer>> actualAssignmentInfo = partitioner.assignmentInfo();
+    Assertions.assertThat(actualAssignmentInfo).isEqualTo(expectedAssignmentInfo);
+
+    Map<Integer, Pair<AtomicLong, Set<RowData>>> partitionResults =
+        runPartitioner(partitioner, numPartitions);
+    validatePartitionResults(expectedAssignmentInfo, partitionResults);
+  }
+
+  @Test
+  public void testNonDividable20PercentClosingFileCost() {
+    MapRangePartitioner partitioner =
+        new MapRangePartitioner(TestFixtures.SCHEMA, SORT_ORDER, mapDataStatistics, 20.0);
+    int numPartitions = 9;
+
+    // target subtask weight is 9 before close file cost factored in.
+    // close file cost is 2 (= 20% * 9) per file.
+    // key weights before and after close file cost factored in
+    // before:     35, 23, 12, 4, 1, 1, 1, 1, 1, 1
+    // close-cost:  8,  6,  4, 2, 2, 2, 2, 2, 2, 2
+    // after:      43, 29, 16, 6, 3, 3, 3, 3, 3, 3
+    // target subtask weight per subtask is 13 (112/9)
+    Map<SortKey, MapRangePartitioner.KeyAssignment> expectedAssignment =
+        ImmutableMap.of(
+            SORT_KEY_0,
+            new MapRangePartitioner.KeyAssignment(
+                new int[] {0, 1, 2, 3}, new long[] {13L, 13L, 13L, 4L}),
```

Comment:
If we want better approximation, then the partition keys should be without the closingFileCost?
- new long[] {11L, 11L, 11L, 2L}

Reply:
here the weights are adjusted with closing file cost

Reply:
But when distributing records, you do not want the consider the closing file costs. You want to have a distribution where to sent the records. The closing file cost comes above the records as a constant

Reply:
I believe this is the same comment as another one above.